sim/midas/src/main/scala/midas/passes/FindClockSources.scala

```diff
@@ -0,0 +1,102 @@
+  def update(renames: RenameMap): Seq[ClockSourceAnnotation] =
+    Seq(this.copy(queryTarget, source.map(s => RTRenamer.exact(renames)(s))))
+}
+
+object FindClockSources extends firrtl.Transform {
+  def inputForm = LowForm
+  def outputForm = LowForm
+  private def getSourceClock(moduleGraphs: Map[String,DiGraph[LogicNode]],
+                             moduleDeps: Map[String, Map[String,String]])
+                            (rT: ReferenceTarget): Option[ReferenceTarget] = {
+    val modulePath   = (OfModule(rT.module) +: rT.path.map(_._2)).reverse
+    val instancePath = (None +: rT.path.map(tuple => Some(tuple._1))).reverse
+
+    def walkModule(currentNode: LogicNode, path: Seq[(Option[Instance], OfModule)]): Option[LogicNode] = {
+      require(path.nonEmpty)
+      val (instOpt, module) :: restOfPath = path
+      val mGraph = moduleGraphs(module.value)
+      // Check if the current node is itself a source. This to handle references to
+      // clocks in the top-level module. All other references must traverse the
+      // graph. Note: The graph is directed from clock-leaves to clock-sources
+      val source = if (mGraph.findSinks(currentNode)) currentNode else {
+        val potentialSources =  mGraph.reachableFrom(currentNode)
+        // FIXME: There should only be a single source here.
+        assert(potentialSources.size >= 1)
+        potentialSources.head
+      }
+      (restOfPath, source) match {
+        // Source is a port on the top level module -> we're done
+        case (Nil, LogicNode(_, None, _)) => Some(source)
+        // Source and sink nodes are the same port on an instance -> unconnected clock port
+        case (_, LogicNode(_, Some(_),_)) if source == currentNode => None
+        // Source is a port on an instance in our current module; prepend it to the path and recurse
+        case (_, LogicNode(port, Some(instName),_)) =>
+          val childModule = moduleDeps(module.value)(instName)
+          walkModule(LogicNode(port), (Some(Instance(instName)), OfModule(childModule)) +: path)
+        // Source is a port but we are not yet at the top; recurse into parent module
+        case (nonEmptyPath, _) => walkModule(LogicNode(source.name, instOpt.map(_.value)), nonEmptyPath)
+      }
+    }
+    val sourceClock = walkModule(LogicNode(rT.ref), instancePath.zip(modulePath))
+    sourceClock.map(sC => rT.moduleTarget.ref(sC.name))
+  }
+
+  def analyze(state: CircuitState, queryTargets: Iterable[ReferenceTarget]): Map[ReferenceTarget, Option[ReferenceTarget]] = {
+    queryTargets.foreach(t => {
+      require(t.component == Nil)
+      require(t.module == t.circuit, s"Queried leaf clock ${t} must provide an absolute instance path")
+    })
+    val c = state.circuit
+    val moduleMap = c.modules.map({m => (m.name,m) }).toMap
+    val iGraph = new InstanceGraph(c).graph
+    val moduleDeps = iGraph.getEdgeMap.map({ case (k,v) => (k.module, (v map { i => (i.name, i.module) }).toMap) }).toMap
```

Comment:
FYI, `InstanceGraph.getChildrenInstances` returns a map from module name (`String`) to instances